### DISCUSSION TITLE
Need advice on DT usage: suppressing a vuln on multiple similar projects

### DISCUSSION THREAD

**Discussion by fmarot** [DependencyTrack/dependency-track#1910] [Q&A]
Hello,
I'm starting to use DT and I really like it. For the moment I have implemented bom creation (including merging java and Javascript sbom) and upload in my CI. My POC goes well for now.

But I have multiple projects all using the same internal framework. I also have multiple versions of the same product that I must track because it is deployed at client's site. In both case, when I analyze a vulnerability and want to set it as "resolved" (and maybe "suppressed"), I'd like to do it on many projects at once as they all share the same problems. Often a non-exploitable vulnerability will not be exploitable whatever the version of the product.
Is there a way to suppress vulnerabilities accross many projects at once ?

I saw projects could be labelled with tags, maybe we can suppress vulnerabilities for all projects having a given tag ? But I did not found how yet.

Regards

**Comment by syalioune** [comment]
Hello @fmarot ,
As far as I know, it is not possible to perform mass auditing with DT.
What you need is probably the approach taken on [dtapac](https://github.com/nscuro/dtapac) by @nscuro

**Reply by fmarot** [reply]
Thanks @syalioune (and sorry for the delay, I had a few days off).
While very interesting, it is a totally new and large process to implement. Do you think proposing a lightweight vulnerability suppressing process accross multiple projects based on tags would interest DependencyTrack owners ?

**Reply by syalioune** [reply]
I honestly can tell but I could recommend to either create an enhancement request with your design thoughts and start the discussion from there or join the slack channel and initiate the discussion with the community (you'll likely have more eyes there).
We've taken the [VEX](https://cyclonedx.org/capabilities/vex/) approach at my company but I personally would be interested to discuss other form of automated auditing.

**Reply by DirkHeinke** [reply]
@syalioune You wrote you use the "VEX approach". I tried to create a VEX file with an affected version range (like [here](https://github.com/CycloneDX/bom-examples/blob/master/VEX/Use-Cases/Case-5/vex.json)) but it seems like DependencyTrack always applies the vex only to the uploaded project. The affects.ref is ignored.

<details>
  <summary>My VEX</summary>

```
{
    "bomFormat": "CycloneDX",
    "specVersion": "1.4",
    "version": 1,
    "metadata": {
        "timestamp": "2023-02-24T08:54:37Z",
        "component": {
            "name": "our.product.name",
            "type": "application",
            "bom-ref": "xxx"
        }
    },
    "vulnerabilities": [
        {
            "id": "GHSA-8v27-2fg9-7h62",
            "source": {
                "name": "GITHUB",
                "url": "https://github.com/advisories"
            },
            "analysis": {
                "state": "false_positive"
            },
            "affects": [
                {
                    "ref": "xxx"
                }
            ]
        }
    ]
}
```
</details>


So how does your "VEX approach" looks like? Do you upload a vex file after each analysis from your CI pipeline?

**Reply by syalioune** [reply]
Hello @DirkHeinke,

With the current DT implementation, a VEX import is always scoped to a defined project (the one that you reference in your VEX upload request).
You can find the current application rules in my comment here https://github.com/DependencyTrack/dependency-track/issues/1872#issuecomment-1254265425 : Affected version range is currently not supported.

> So how does your "VEX approach" looks like? Do you upload a vex file after each analysis from your CI pipeline?

Yeah, we've built a custom CLI around DT REST API. That CLI has an upload command that takes a SBOM parameter and optionally a VEX  parameter. If a VEX is supplied to that command, the CLI upload the SBOM, wait for the analysis to complete and then upload the VEX to the same project. All those actions happens in a Gitlab CI pipeline. We recommend that each of our company projects manage a VEX file at the root of their git repository so that it is versioned along with the related source code.

**Reply by DirkHeinke** [reply]
Hi @syalioune 

thank you for your reply. I guess your CLI is not public somewhere? For my use case I'll try to use the clone API to copy over suppressions from the old version.

**Reply by syalioune** [reply]
> thank you for your reply. I guess your CLI is not public somewhere?

Nope, it is an internal project 😢 

> For my use case I'll try to use the clone API to copy over suppressions from the old version

That should work as long as your clone request object contains `includeAuditHistory: true`
